File: src/mouse.rs
use std::{
    fs::{self, read_dir, File, OpenOptions},
    io::Write,
    os::unix::fs::FileTypeExt,
    time::{SystemTime, UNIX_EPOCH},
};

use glam::{IVec2, Vec2};

use crate::config::DEBUG_WITHOUT_MOUSE;

#[derive(Clone, Debug, PartialEq)]
pub enum MouseStatus {
    Working(String),
    Disconnected,
    PermissionsRequired,
    NoMouseFound,
}

#[derive(Debug, Clone, Copy)]
struct Timeval {
    seconds: u64,
    microseconds: u64,
}

#[derive(Debug, Clone, Copy)]
struct InputEvent {
    time: Timeval,
    event_type: u16,
    code: u16,
    value: i32,
}

impl InputEvent {
    fn bytes(&self) -> Vec<u8> {
        let mut bytes: Vec<u8> = Vec::with_capacity(24);

        bytes.extend(&self.time.seconds.to_le_bytes());
        bytes.extend(&self.time.microseconds.to_le_bytes());

        bytes.extend(&self.event_type.to_le_bytes());
        bytes.extend(&self.code.to_le_bytes());
        bytes.extend(&self.value.to_le_bytes());

        bytes
    }
}

const EV_SYN: u16 = 0x00;
const EV_REL: u16 = 0x02;
const SYN_REPORT: u16 = 0x00;
const AXIS_X: u16 = 0x00;
const AXIS_Y: u16 = 0x01;
const BTN_LEFT: u16 = 0x110;
const BTN_RIGHT: u16 = 0x111;

pub fn open_mouse() -> (File, MouseStatus) {
    if DEBUG_WITHOUT_MOUSE {
        let file = OpenOptions::new().write(true).open("/dev/null").unwrap();
        return (file, MouseStatus::Working(String::from("/dev/null")));
    }
    for file in read_dir("/dev/input").unwrap() {
        let entry = file.unwrap();
        if !entry.file_type().unwrap().is_char_device() {
            continue;
        }
        let name = entry.file_name().into_string().unwrap();
        if !name.starts_with("event") {
            continue;
        }
        // get device info from /sys/class/input
        let keys: Vec<u32> =
            fs::read_to_string(format!("/sys/class/input/{}/device/capabilities/key", name))
                .unwrap()
                .split_whitespace() // Handle multiple hex numbers
                .filter_map(|hex| u32::from_str_radix(hex, 16).ok()) // Parse each hex number
                .flat_map(decompose_bits) // Decompose into individual bits
                .collect();
        let rel: Vec<u32> =
            fs::read_to_string(format!("/sys/class/input/{}/device/capabilities/rel", name))
                .unwrap()
                .split_whitespace() // Handle multiple hex numbers
                .filter_map(|hex| u32::from_str_radix(hex, 16).ok()) // Parse each hex number
                .flat_map(decompose_bits) // Decompose into individual bits
                .collect();
        if !rel.contains(&(AXIS_X as u32))
            && !rel.contains(&(AXIS_Y as u32))
            && !keys.contains(&(BTN_LEFT as u32))
            && !keys.contains(&(BTN_RIGHT as u32))
        {
            continue;
        }
        let device_name =
            fs::read_to_string(format!("/sys/class/input/{}/device/name", name)).unwrap();

        let path = format!("/dev/input/{}", name);
        let file = OpenOptions::new().write(true).open(path);
        match file {
            Ok(file) => return (file, MouseStatus::Working(device_name)),
            Err(_) => {
                println!("please add your user to the input group or execute with sudo.");
                println!(
                    "without this, mouse movements will be written to /dev/null and discarded."
                );
                let file = OpenOptions::new().write(true).open("/dev/null").unwrap();
                return (file, MouseStatus::PermissionsRequired);
            }
        }
    }

    let file = OpenOptions::new().write(true).open("/dev/null").unwrap();
    (file, MouseStatus::NoMouseFound)
}

pub fn move_mouse(mouse: &mut File, coords: Vec2) {
    let coords = IVec2::new(coords.x as i32, coords.y as i32);
    if DEBUG_WITHOUT_MOUSE {
        println!("moving mouse: ({} / {})", coords.x, coords.y);
        return;
    }

    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    let time = Timeval {
        seconds: now.as_secs(),
        microseconds: now.subsec_micros() as u64,
    };

    let x = InputEvent {
        time,
        event_type: EV_REL,
        code: AXIS_X,
        value: coords.x,
    };

    let y = InputEvent {
        time,
        event_type: EV_REL,
        code: AXIS_Y,
        value: coords.y,
    };

    let syn = InputEvent {
        time,
        event_type: EV_SYN,
        code: SYN_REPORT,
        value: 0,
    };

    mouse.write_all(&x.bytes()).unwrap();
    mouse.write_all(&syn.bytes()).unwrap();

    mouse.write_all(&y.bytes()).unwrap();
    mouse.write_all(&syn.bytes()).unwrap();
}

const SYN: InputEvent = InputEvent {
    time: Timeval {
        seconds: 0,
        microseconds: 0,
    },
    event_type: EV_SYN,
    code: SYN_REPORT,
    value: 0,
};

pub fn mouse_valid(mouse: &mut File) -> bool {
    if mouse.write_all(&SYN.bytes()).is_ok() {
        return true;
    }
    false
}

fn decompose_bits(bitmask: u32) -> Vec<u32> {
    (0..32)
        .filter(|bit| (bitmask & (1 << bit)) != 0) // Check if the bit is set
        .collect()
}

File: src/mouse.cpp
#include "mouse.hpp"

#include <fcntl.h>
#include <linux/input-event-codes.h>
#include <linux/input.h>
#include <unistd.h>

#include <cerrno>
#include <filesystem>
#include <fstream>
#include <mithril/logging.hpp>
#include <mithril/types.hpp>

i32 mouse = 0;

std::vector<bool> HexToReversedBinary(char hex_char) {
    int value = std::isdigit(hex_char) ? hex_char - '0' :
                (std::isxdigit(hex_char) && std::islower(hex_char)) ? hex_char - 'a' + 10 :
                hex_char - 'A' + 10;

    std::vector<bool> bits(4);
    for (int i = 0; i < 4; ++i) {
        bits[i] = (value >> i) & 1;
    }
    return bits;
}

std::vector<bool> DecodeCapabilities(const std::string &filename) {
    std::ifstream file(filename);
    if (!file) {
        logging::Warning("cannot open mouse capability device file: {}", filename);
        return {};
    }

    std::string content;
    std::getline(file, content);

    std::vector<bool> binary_out;
    int hex_count = 0;

    for (auto it = content.rbegin(); it != content.rend(); ++it) {
        char c = *it;
        if (c == ' ') {
            binary_out.insert(binary_out.end(), 4 * (16 - hex_count), false);
            hex_count = 0;
        } else if (std::isxdigit(c)) {
            const auto bits = HexToReversedBinary(c);
            binary_out.insert(binary_out.end(), bits.begin(), bits.end());
            hex_count++;
        }
    }

    if (hex_count > 0 && hex_count < 16) {
        binary_out.insert(binary_out.end(), 4 * (16 - hex_count), false);
    }

    return binary_out;
}

void MouseInit() {
    for (const auto &entry : std::filesystem::directory_iterator("/dev/input")) {
        if (!entry.is_character_file()) continue;

        const auto &path = entry.path();
        const std::string name = path.filename().string();
        if (name.rfind("event", 0) != 0) continue;

        const std::string rel_caps_path = "/sys/class/input/" + name + "/device/capabilities/rel";
        const auto rel_caps = DecodeCapabilities(rel_caps_path);
        if (rel_caps.size() <= REL_Y || !rel_caps[REL_X] || !rel_caps[REL_Y]) continue;

        const std::string key_caps_path = "/sys/class/input/" + name + "/device/capabilities/key";
        const auto key_caps = DecodeCapabilities(key_caps_path);
        if (key_caps.size() <= BTN_LEFT || !key_caps[BTN_LEFT]) continue;

        const std::string name_path = "/sys/class/input/" + name + "/device/name";
        std::ifstream name_file(name_path);
        if (!name_file) continue;

        std::string device_name;
        std::getline(name_file, device_name);

        mouse = open(path.c_str(), O_WRONLY);
        if (mouse < 0) {
            logging::Error("could not open mouse: {}", path.string());
            if (errno == EACCES) {
                if (const char* user = getlogin(); user) {
                    logging::Error("user '{}' is not in input group. Try: sudo usermod -aG input {}", user, user);
                }
            } else {
                logging::Error("error code: {}", errno);
            }
            std::exit(1);
        }

        logging::Info("found mouse: {} ({})", device_name, name);
        return;
    }

    mouse = open("/dev/null", O_WRONLY);
    logging::Warning("no mouse device found, using /dev/null");
    std::exit(1);
}

void MouseQuit() {
    if (mouse > 0) {
        close(mouse);
    }
}

void MouseMove(const glm::ivec2 &coords) {
    logging::Debug("mouse move: {}/{}", coords.x, coords.y);

    input_event ev {
        .type = EV_REL,
        .code = REL_X,
        .value = coords.x
    };
    write(mouse, &ev, sizeof(ev));

    ev.code = REL_Y;
    ev.value = coords.y;
    write(mouse, &ev, sizeof(ev));

    ev.type = EV_SYN;
    ev.code = SYN_REPORT;
    ev.value = 0;
    if (write(mouse, &ev, sizeof(ev)) <= 0) {
        logging::Warning("mouse write failed");
    }
}

void MouseLeftPress() {
    logging::Debug("pressed left mouse button");
    input_event ev {
        .type = EV_KEY,
        .code = BTN_LEFT,
        .value = 1
    };
    write(mouse, &ev, sizeof(ev));

    ev.type = EV_SYN;
    ev.code = SYN_REPORT;
    ev.value = 0;
    if (write(mouse, &ev, sizeof(ev)) <= 0) {
        logging::Warning("mouse write failed");
    }
}

void MouseLeftRelease() {
    logging::Debug("released left mouse button");
    input_event ev {
        .type = EV_KEY,
        .code = BTN_LEFT,
        .value = 0
    };
    write(mouse, &ev, sizeof(ev));

    ev.type = EV_SYN;
    ev.code = SYN_REPORT;
    ev.value = 0;
    if (write(mouse, &ev, sizeof(ev)) <= 0) {
        logging::Warning("mouse write failed");
    }
}

bool MouseValid() {
    input_event ev {
        .time = {.tv_sec = 0, .tv_usec = 0},
        .type = EV_SYN,
        .code = SYN_REPORT,
        .value = 0
    };
    return write(mouse, &ev, sizeof(ev)) > 0;
}
